### ISSUE TITLE
Migrations always generated for custom type with DBAL 4

### ISSUE THREAD

**Issue by michnovka** [doctrine/migrations#1441]
### Bug Report

|    Q        |   A
|------------ | ------
| BC Break    | no
| Migrations version     | 3.8.0
| Migrations bundle | 3.3.1
| ORM     | 3.2.1
| DBAL     | 4.0.4
| Symfony | 7.1.2

#### Summary

I have issue with DBAL 4 and custom type, the migrations keep getting generated again and again. Furthermore, the down migration looks totally bogus. This is possibly related to #1435 

I know that DBAL 4 dropped requiresSqlHint (in https://github.com/doctrine/dbal/pull/5107 , afterwards some issues were found and fixed - https://github.com/doctrine/dbal/issues/6257 )

So when I am using custom type, I expect the first migration diff to drop the `DC2Type` comments. However my tables have these fields already dropped and yet the migration is being generated.

```php
enum ActionType: string
{
    case SUBMIT = 'submit';
    case CANCEL = 'cancel';
}

class ActionTypeType extends Type
{
    public function getSQLDeclaration(array $column, AbstractPlatform $platform): string
    {
        $class = ActionType::class;

        $values = [];

        foreach ($class::cases() as $val) {
            $values[] = "'{$val->value}'";
        }

        return "enum(" . implode(", ", $values) . ")";
    }

    public function convertToDatabaseValue($value, AbstractPlatform $platform): mixed
    {
        $class = ActionType::class;
        if ($value !== null && !($value instanceof BackedEnum)) {
            $value = $class::tryFrom($value);
        }else{
            return null;
        }

        return $value->value;
    }

    public function convertToPHPValue($value, AbstractPlatform $platform): ?BackedEnum
    {
        if ((!is_int($value)) && !is_string($value)) {
            return null;
        }

        return $class::tryFrom($value);
    }
}
```

I then have my entity as

```php
#[ORM\Entity(repositoryClass: ActionRepository::class)]
#[ORM\Table(name: 'actions')]
class Action
{
    #[ORM\Id]
    private string $name;

    #[ORM\Column(type: ActionType::class, nullable: false)]
    private ActionType $type;
}
```

and in `Kernel.php` I set up type mapping inside `Kernel::process()`:

```php
$typesDefinition[ActionType::class] = ['class' => ActionTypeType::class];
$container->setParameter('doctrine.dbal.connection_factory.types', $typesDefinition);
```

Now I know that the types are assigned correctly, as migrations generate up like this:

```php
    public function up(Schema $schema): void
    {
        $this->addSql('ALTER TABLE actions CHANGE type type enum(\'submit\', \'cancel\') NOT NULL');
    }
```
but it is generated ALWAYS.

and the `down()` migration looks even weirder:

```php
    public function down(Schema $schema): void
    {
        $this->addSql('ALTER TABLE actions CHANGE type type VARCHAR(0) NOT NULL');
    }
```

Everything works fine with DBAL 3, which uses SQL comments

**Comment by michnovka** [comment]
@greg0ire I am tagging you as you authored some of the relevant changes and I know you have deep insight into removing comments in DBAL 4.

I was digging into the code, the issue lies here:

So `Doctrine\Migrations\Generator\DiffGenerator::generate()` calls `createFromSchema` to do comparison of old vs new schema. This calls `Doctrine\DBAL\Schema\MySQLSchemaManager` (which inherits `AbstractSchemaManager`) and specifically it fails on function `_getPortableTableColumnDefinition()`. Where the array `$tableColumn` which this function gets is

```
 array:11 [
  "table_name" => "actions"
  "field" => "type"
  "type" => "enum('submit','cancel')"
  "null" => "NO"
  "key" => ""
  "default" => null
  "extra" => ""
  "comment" => ""
  "characterset" => "utf8mb4"
  "collation" => "utf8mb4_general_ci"
  "name" => ""
]
```

This function then calls `$type = $this->platform->getDoctrineTypeMapping($dbType)` and since enum, or whatever custom type is not recognized, it defaults to `StringType`.

Now check DBAL 3 code:
https://github.com/doctrine/dbal/blob/893417fee2bc5a94a10a2010ae83cab927e21df3/src/Schema/MySQLSchemaManager.php#L189-L192

This uses the comment to deduce the proper type and override whatever type is in DB. Without this now, DBAL considers any unknown DB type to be `StringType` and I have no way how to tell it its not a string type. Hence the migration will always consider old schema column to be `string`.

Now the new schema uses the `DiffGenerator::createToSchema()` which does not use `Doctrine\DBAL\Schema\AbstractSchemaManager` but instead `Doctrine\Migrations\Provider\OrmSchemaProvider` and this uses under the hood `EntityManager` which of course identifies the column type properly as `ActionTypeType` or whatever custom type you use.

So, since we removed comments in DBAL 4, how can the `fromSchema` be generated properly? This uses only DB values, it does not care about type definition in Entity code (thats relevant for `toSchema` only).

**Comment by greg0ire** [comment]
Please upgrade the ORM to 3.8.5: https://github.com/doctrine/dbal/pull/6423

EDIT: I shouldn't answer so hastily

**Comment by michnovka** [comment]
@greg0ire I am using DBAL 4.0.4 which is based on 3.8.6 that includes the linked fix. ORM is at 3.2.1 and there is no newer version. And the issue is ONLY with DBAL 4 not DBAL 3

**Comment by berkut1** [comment]
@michnovka It seems that when Doctrine reads the schema of your database, the `getSQLDeclaration` function is not being called, which is why it cannot recognize the type. There might be a bug in https://github.com/doctrine/dbal/blob/4.0.x/src/Platforms/AbstractMySQLPlatform.php

~~Try debugging this function to determine at which stage the type is being lost:
https://github.com/doctrine/dbal/blob/90424473eb144659a89fb1b5c9bca37297085351/src/Platforms/AbstractPlatform.php#L2190~~
Probably, this is a mistaken assumption. Here, two schemas columns are being compared. If initially it was read as StringType, then the error is somewhere earlier.

**Comment by michnovka** [comment]
@berkut1 this is indeed a mistaken assumption imo, as getSQLDeclaration not being called is a consequence of not having the proper type. The type has to be known in order for that function to be called. And it is called, just for the new schema. 

The issue is that oldSchema is created based on database ONLY. It ignores PHP code. And thats good, as the php code defines the new schema, if it took the PHP code as a hint of the type, then migration has no point, as it would be the same type always.

I honestly see no other way than to somehow store in DB the name of the custom type which was used. Which is exactly what the DB comment was used for. WDYT?

**Comment by berkut1** [comment]
@michnovka 
No, DC2Type comments have been useless since around DBAL 3.x. Doctrine can determine custom types on its own if they are properly registered. You need to find out why, in your case, this is not happening for `Enum` and where exactly the error is in the code.

**Comment by michnovka** [comment]
@berkut1 how are they useless when even in 3.9.x there is this code:

https://github.com/doctrine/dbal/blob/893417fee2bc5a94a10a2010ae83cab927e21df3/src/Schema/MySQLSchemaManager.php#L189-L192 ?

And this is where it fails, this is where in 3.8.x version it assigned properly custom type, but in 4.0.x it assigns StringType (as the code assigning type based on comment is missing)

And again, think about what I observed above - the oldSchema is created based on DB data ONLY. not PHP. Only DB. So if DB has no indication about the type, how can it be properly determined for nonstandard types?
newSchema uses PHP code to determine types. And thats good.

> Doctrine can determine custom types on its own if they are properly registered.

This has no sense for migrations. If I had `customType1` field in table, and I changed it to `customType2` field, then did migration, how would doctrine deduce it was before `customType1`? It is no longer defined and registered. PHP code always reflects only the newSchema.

**Comment by berkut1** [comment]
DBAL 3.x should have backward compatibility, that why it can works with DC2Types and without. there was added this https://github.com/doctrine/dbal/blob/893417fee2bc5a94a10a2010ae83cab927e21df3/src/Platforms/AbstractPlatform.php#L108
that people can smoothly migrate from DBAL3 to DBAL4

**Comment by michnovka** [comment]
With DBAL 3.8.X and `doctrine.dbal.disable_type_comments: true` the issue is also present.

And of course it is. Read again[ where and why it happens](https://github.com/doctrine/migrations/issues/1441#issuecomment-2206338070), I explained exactly where the issue lies. oldSchema is created based on DB details only. If we dont use comments (either as they are removed in DBAL 4, or if we disable them manually in DBAL 3) then oldSchema will have no idea what type was used before.

Thanks for your interest and will to help.

**Comment by berkut1** [comment]
Here, 
https://github.com/doctrine/dbal/blob/90424473eb144659a89fb1b5c9bca37297085351/src/Schema/MySQLSchemaManager.php#L135
and
https://github.com/doctrine/dbal/blob/90424473eb144659a89fb1b5c9bca37297085351/src/Schema/MySQLSchemaManager.php#L218
Doctrine is trying to map your registered custom types to your database. What happens when it tries to map the Enum type in your case?

UPD:
You also can check if your type registered, here:
![image](https://github.com/doctrine/migrations/assets/13355990/fb1a1d73-7976-4dbe-a499-510e09a7dd37)

**Comment by michnovka** [comment]
> https://github.com/doctrine/dbal/blob/90424473eb144659a89fb1b5c9bca37297085351/src/Schema/MySQLSchemaManager.php#L135

```
dump($type, $dbType);
```
gives
```
1 "string"
2 "enum"
```

and

> https://github.com/doctrine/dbal/blob/90424473eb144659a89fb1b5c9bca37297085351/src/Schema/MySQLSchemaManager.php#L218
```
dump($column);
```
gives
```
Doctrine\DBAL\Schema\Column {#926
  #_name: "type"
  #_namespace: null
  #_quoted: false
  #_type: Doctrine\DBAL\Types\StringType {#636}
  #_length: 0
  #_precision: null
  #_scale: 0
  #_unsigned: false
  #_fixed: false
  #_notnull: true
  #_default: null
  #_autoincrement: false
  #_platformOptions: []
  #_columnDefinition: null
  #_comment: ""
}
```

UPD:
I checked and the `typeRegistry` does contain my custom type `ActionTypeType`

**Comment by greg0ire** [comment]
@michnovka it's unclear to me whether you know what platform-aware comparison is. In case you don't, since DBAL 3.2.0, platform-aware comparison allows to compare the generated SQL instead of comparing schemas. This means that 2 different schemas could result in an empty diff, if they result in the same SQL being generated. More on this here: https://www.doctrine-project.org/2021/11/26/dbal-3.2.0.html

Sorry if you already know this, it's genuinely hard to tell for me.

**Comment by michnovka** [comment]
@greg0ire yes, I get this. 

The problem is that when the comparator tries to compare the schemas, it is working with oldSchema and newSchema. oldSchema is `Schema` instance which is created and populated with data using `Doctrine\DBAL\Schema\MySQLSchemaManager` and it considers my column to be `StringType` (cos DB column definition is `enum(...)` which it doesnt understand, so it says, its `StringType` - this can be seen in my debug output in comment https://github.com/doctrine/migrations/issues/1441#issuecomment-2207273101). 

This means that when it tries to compare SQL generated between old and new schemas, it calls `StringType::getSQLDeclaration` output instead of the `ActionTypeType::getSQLDeclaration` for the old schema.

And this is expected. As if in DB there is no info about which custom type the column is, how can the `MySQLSchemaManager` assume its `ActionTypeType` and then gets its SQL from `ActionTypeType::getSQLDeclaration` to compare with new schema.

Now the new schema is another `Schema` instance, but this one is created using `Doctrine\Migrations\Provider\OrmSchemaProvider` and this calls internally `EntityManager` and that uses PHP code and properly considers the column `ActionTypeType`. Then new schema's SQL is the proper one, while the old ones is nonsense, as it thinks its a `StringType`.

**Comment by greg0ire** [comment]
Ah I get it now, thanks for explaining it again. I think this means the issue is that the DBAL does not understand `enum(…)`, and cannot find a type that will produce the same DDL. I don't see a way around that. I now better get the point of https://github.com/doctrine/dbal/pull/6444, which I now realize was about the same thing.

**Comment by michnovka** [comment]
yes, it all comes to a simple problem - the oldSchema is generated using ONLY DB introspection. And if DB contains no info about which type was used, then it cannot call proper `getSQLDeclaration` function.

And no, the problem does not simply limit to enum. I can make an example with another custom type which will have the same issue without enums. ALL custom types have this issue in fact, i.e. all custom types with custom `getSQLDeclaration` which differs from those provided by DBAL package.

**Comment by berkut1** [comment]
Here, Doctrine is reading your database. Does it also read the type as a string here?
https://github.com/doctrine/dbal/blob/90424473eb144659a89fb1b5c9bca37297085351/src/Schema/AbstractSchemaManager.php#L195

leads to https://github.com/doctrine/dbal/blob/90424473eb144659a89fb1b5c9bca37297085351/src/Schema/MySQLSchemaManager.php#L344

**Comment by greg0ire** [comment]
This looks related: https://github.com/doctrine/dbal/issues/5308#issuecomment-1058391601

**Comment by michnovka** [comment]
> Here, Doctrine is reading your database. Does it also read the type as a string here? https://github.com/doctrine/dbal/blob/90424473eb144659a89fb1b5c9bca37297085351/src/Schema/AbstractSchemaManager.php#L195

No, here it fetches it from DB like this:
```
   1 => array:10 [
      "TABLE_NAME" => "actions"
      "field" => "type"
      "type" => "enum('submit','cancel')"
      "null" => "NO"
      "key" => ""
      "default" => null
      "EXTRA" => ""
      "comment" => ""
      "characterset" => "utf8mb4"
      "collation" => "utf8mb4_general_ci"
    ]
```

But this is just one step before `_getPortableTableColumnDefinition` is called which is where the `$this->platform->getDoctrineTypeMapping($dbType);` is called, and which we already addressed and debugged in  https://github.com/doctrine/migrations/issues/1441#issuecomment-2207273101

And @greg0ire this is where the issue lies: 
https://github.com/doctrine/dbal/blob/90424473eb144659a89fb1b5c9bca37297085351/src/Schema/MySQLSchemaManager.php#L218
```php
        $column = new Column($tableColumn['field'], Type::getType($type), $options);
```

specifically the `Type::getType($type)`. Without comments, there is no way at all for the old type to be "guessed" correctly. If I use custom type which has SQL declaration always as int(3), (lets call it `CustomInt3Type`), then the `Column` instance will always be deduced as `IntegerType` and for comparison `IntegerType::getSQLDeclaration` will be called to be compared against newSchema's `CustomInt3Type::getSQLDeclaration`. There is just no way for it to know that this int(3) is actually `CustomInt3Type` and not `IntegerType`

**Comment by berkut1** [comment]
Ah, I understand now. It calls `Type::getType` with `enum` and cannot find such a type because it doesn't exist.
In Symfony, you can work around this by registering another type to let Doctrine know about it: https://symfony.com/doc/current/doctrine/dbal.html#registering-custom-mapping-types-in-the-schematool (this instruction correct for DBAL3)
In your case, for DBAL 4 with Symfony, you need to do it like this:
```yaml
        mapping_types:
            enum: ancestor_enum 
        types:
            ancestor_enum : 'App\Model\AncestorEnumType' 
```
Unfortunately, I can't help with how to do this without Symfony.

UPD:
In the example, I specified the type as `ancestor`, assuming that you can register it in the mapping (`mapping_types`) only once. For all your other `Enums`, it will be enough to inherit from it and register only as a custom type (`types`). Otherwise, you would have to duplicate entries for all Enums in `mapping_types`.

**Comment by michnovka** [comment]
@greg0ire as a temporary workaround, how can I tell migration to ignore certain columns? Thanks!

**Comment by greg0ire** [comment]
Maybe you can use asset filtering, but I don't think it works at the column level.

**Comment by michnovka** [comment]
So the issue is this:

1. we get table columns in raw format from DB using `DBAL\Schema\AbstractSchemaManager::fetchTableColumnsByTable()` This is an array like 
```
   1 => array:10 [
      "TABLE_NAME" => "actions"
      "field" => "type"
      "type" => "enum('submit','cancel')"
      "null" => "NO"
      "key" => ""
      "default" => null
      "EXTRA" => ""
      "comment" => ""
      "characterset" => "utf8mb4"
      "collation" => "utf8mb4_general_ci"
    ]
```
2. This is converted to `DBAL\Schema\Column` instance, which has incorrect type (as DBAL has no idea what custom type is to be used without comments, so it guesses best match from default types)
3. then `Column::getType()::getSQLDeclaration()` is called to obtain incorrect SQL

And on the new schema, we get correct type for column and call its `getSQLDeclaration()`

These do not match.

---

Proposed solution:

Why not leverage `Column::$_columnDefinition` for the old schema? Lets populate this field inside `AbstractSchemaManager::_getPortableTableColumnDefinition`. 

I dont understand why `AbstractPlatform::columnsEqual()` unsets this. But if we dont unset it and specify it explicitly for old schema, the comparison would be done properly

@greg0ire WDYT?

**Comment by greg0ire** [comment]
Sounds good, feel free to give it a try, and see if anything breaks 👍

**Comment by michnovka** [comment]
I tried to go this way and I am afraid this would create tons of unexpected issues. There are specific test cases that ensure that 

> Schema diff is empty, since only `columnDefinition` changed from `null` (not detected) to a defined one

`AbstractComparatorTestCase::testWillNotProduceSchemaDiffOnTableWithAddedCustomSchemaDefinition`

so for whatever reason this is considered important. Even the comment in the `AbstractPlatform::columsEqual` talks about this specifically to ignore the column definition.

So maybe a solution would be to add comments back so that the type can be deduced properly?

**Comment by greg0ire** [comment]
I don't think that's the direction we want to go, no. Re-reading [the cookbook you mentioned](https://www.doctrine-project.org/projects/doctrine-orm/en/3.2/cookbook/mysql-enums.html) I see

> In this case however Schema-Tool update will have a hard time not to request changes for this column on each call.

It seems like that might have been fixed in https://github.com/doctrine/dbal/pull/5224, which might make solution 1 the way to go here.

**Comment by berkut1** [comment]
@michnovka 
I don't know why you ignored my solution https://github.com/doctrine/migrations/issues/1441#issuecomment-2207383215. For reference, I had a very similar problem, not with `enum`, but with the `inet` type, which is unknown to Doctrine. It tried, as in your case, to convert it to `VARCHAR` in the `down()` method and constantly generated diff migrations.

As I mentioned in my comment on the instructions, I solved the problem for myself in this way:
```yaml
doctrine:
    dbal:
        mapping_types:
            inet: my_inet

        types:
            my_inet: { class: 'App\Model\InetType' }
```

I want to clarify, as stated in the instructions at https://www.doctrine-project.org/projects/doctrine-orm/en/3.2/cookbook/mysql-enums.html, it is recommended to do it like this
```php
$conn->getDatabasePlatform()->registerDoctrineTypeMapping('enum', 'string');
```
but this solution works only for DBAL3. For DBAL4, you need to refer not to `string` but to your newly registered custom type, as I indicated in my previous comment.

However, if you haven't tried this solution at all, you can first refer to `string`; it might work with `enum`.

**Comment by michnovka** [comment]
@berkut1 Thanks for the suggestion, but this is not a solution for my case.

It works in your case when you have 1:1 mapping between DB type `INET` and your class `MY_INET`. So your code is 100% certain, when it finds `INET` type in DB to use that specific type.

But in my case, every `ENUM` is different. The example I provided is `ENUM('submit','cancel')` and it uses custom type `ActionTypeType extends AbstractEnumType`. `AbstractEnumClass` is my base class for enums, and I have many others. There is e.g. `ColorEnumType extends AbstractEnumType` which creates `ENUM('black','red')`. So DB `ENUM` type maps to multiple PHP `Type` classes.

**Comment by berkut1** [comment]
@michnovka 

Yes, I understand, which is why I initially suggested creating an abstract Enum class and registering it, then inheriting from it for custom types. 
If I understand correctly how Doctrine works with types, when comparing schemas, it only compares the types of SQL, and the comparison of `ColumnDeclarationSQL` happens only at the final stage: https://github.com/doctrine/dbal/blob/90424473eb144659a89fb1b5c9bca37297085351/src/Platforms/AbstractPlatform.php#L2199-L2204

In other words, you need to make Doctrine understand that it has a base Enum class for comparison, with which it will compare all the descendants of this class.
This test for DBAL3 with disabled type comments, show the logic:
https://github.com/doctrine/dbal/blob/893417fee2bc5a94a10a2010ae83cab927e21df3/tests/Platforms/AbstractPlatformTestCase.php#L1448

**Comment by michnovka** [comment]
@berkut1 this does not work for the reason I mentioned. The same way that for doctrine `VARCHAR(16)` and `VARCHAR(65)` are both the same `StringType`, it considers `ENUM('black','red')` and `ENUM('fast','slow')` as the same type. Since the type has no idea about which enum to use as a base.

So what I can do, is map `ENUM(WHATEVER)` to some "abstract" Enum class (notice the quotes around abstract, as the class cannot be abstract, as it has to be instantiated).

But then, how can this "abstract" enum class generate valid SQL code in `getSQLDeclaration`, and how can it cast properly to the actual specific PHP enum in `convertToPHPValue` ? So the approach I took, is I do actually have an `AbstractEnumType` class and I extend this for every single PHP enum I want to use as MySQL ENUM. This way every column that is mapped to this type is actually an instance of some child of `AbstractEnumType` but it is a different child for every enum I use, therefore it performs proper casting and can create the appropriate SQL declaration.

And this is where we come to the issue which I am not sure how to address with some easy fix. `ENUM('black','red')` and `ENUM('fast','slow')` should be treated as different types.

The only solution is to compare the SQL read from `INFORMATION_SCHEMA` with the SQL generated for a new schema.

Now we do:
1. read `INFORMATION_SCHEMA` from DB and populate old schema's `Column` using **DBAL's** `SchemaManager`<- this is where it fails, as the Type field of old schema is not 100% correct with custom types
2. populate new schema's `Column` using **Migration's** `DiffGenerator::createToSchema` <- this is correct, as it reads types properly from attributes
3. We compare the 2 columns <- they always differ as the old schema cannot know the proper type to use.

I tried to implement https://github.com/doctrine/migrations/issues/1441#issuecomment-2208783407 leveraging `Column::$_columnDefinition` but this is not OK, it creates tons of failing tests because this is not what the field was meant for.

**Comment by michnovka** [comment]
@greg0ire Idea for workaround:

How about we implement an attribute `#[Doctrine\Migration\Mapping\IgnoreInMigrations]` and if we use it on a type, it will ignore the column, or potentially whole Entity? If new schema has a column/entity with this attribute, it will simply not care if the old schema has it/does not have it/has it different. Simply skip any check.

This is different than DBAL's `schema_filter`, since that one filters out tables that are present in the database.

Migration bundle can then show a warning that some columns/tables were ignored. And its up to the developer to manage such fields in migrations manually.

**Comment by PowerKiKi** [comment]
I am having the exact same issue as @michnovka. I have a bunch of different PHP enums, each of them is used to declare a DBAL type enum via inheritance of a custom `PhpEnumType` DBAL type.  That parent will use `ENUM` type in DB, instead of the `VARCHAR` used by [the column attribute `enumType`](https://www.doctrine-project.org/projects/doctrine-orm/en/3.2/reference/basic-mapping.html#property-mapping). So it's basically https://knplabs.com/en/blog/how-to-map-a-php-enum-with-doctrine-in-a-symfony-project/

But now migrations are always re-generated for those columns :-/

@michnovka, the workaround you suggest seems like quite a big regression feature-wise. Having auto-managed enums is really great. And there is no reason that Doctrine should not treat ENUM differently than any other types. It should be supported as well as other types. And if not out-of-the-box, at least Doctrine should let us the possibility to implement them with reasonable effort.

@greg0ire I get it that DBAL does not require comments anymore for its out of the box behavior. Doctrine is improving, that's good for all of us. But it does not mean that comments don't have any real use outside of DBAL internals, to solve real use-cases.

The fact that the removal of comment broke the [cookbook to use `ENUM`](https://www.doctrine-project.org/projects/doctrine-orm/en/3.2/cookbook/mysql-enums.html#solution-2-defining-a-type) in DB is very telling. You suggest to fallback on the [cookbook that use `VARCHAR`](https://www.doctrine-project.org/projects/doctrine-orm/en/3.2/cookbook/mysql-enums.html#solution-1-mapping-to-varchars) instead. It might work for some, but it might also be unacceptable for some project to change all column type from `ENUM` to `VARCHAR`.

I think this whole thing boils down to: how can Doctrine supports `ENUM` as a first-class citizen, and not what seems to only be an afterthought ?

_PS: but https://github.com/doctrine/dbal/issues/5308#issuecomment-1058391601 suggests that there is a broader issue that goes beyond `ENUM` too..._

**Comment by greg0ire** [comment]
>to change all column type from ENUM to VARCHAR.

Can you clarify exactly what you mean by that?

> how can Doctrine supports ENUM as a first-class citizen, and not what seems to only be an afterthought ?

[It would seem like it cannot](https://github.com/doctrine/dbal/issues/6465)

**Comment by PowerKiKi** [comment]
> > to change all column type from ENUM to VARCHAR.
> 
> Can you clarify exactly what you mean by that?

What I mean that is that before DBAL 4.0, the [cookbook solution 2](https://www.doctrine-project.org/projects/doctrine-orm/en/3.2/cookbook/mysql-enums.html#solution-2-defining-a-type) allowed to have a DB type of `ENUM`.

But now it will re-generate migrations everytime. So it appears the only way to avoid that, is to use the built-in support for PHP enum, via either the `enumType` attribute, or the implicit typed field mapper, as shown below. Because I don't think the cookbook solution 1 can work with PHP enum without unreasonable boilerplate. But both of those solutions will change the DB colum from `ENUM` to `VARCHAR`, and thus losing the [advantages of DB enum](https://github.com/doctrine/dbal/issues/6465#issuecomment-2213955732). And our DB will become a mess of non-validated data.

Given this entity:

```php
#[ORM\Entity]
class Task
{
    #[ORM\Column(type: 'Priority')]
    private Priority $priorityViaCustomType;

    #[ORM\Column(enumType: Priority::class)]
    private Priority $priorityViaEnumType;

    #[ORM\Column]
    private Priority $priorityViaTypedFieldMapper;
}
```

I will get the following SQL for MariaDB:

```sql
ALTER TABLE task
ADD priority_via_custom_type ENUM('low', 'normal', 'high') NOT NULL COMMENT '(DC2Type:Priority)', -- DBAL 4.0 won't have the comment anymore, so it will generate migrations everytime 😭 
ADD priority_via_enum_type VARCHAR(255) NOT NULL, -- Not validated by DB 😭
ADD priority_via_typed_field_mapper VARCHAR(255) NOT NULL; -- Not validated by DB 😭
```

**Comment by berkut1** [comment]
@PowerKiKi If your database has dependencies on the ENUM attribute type, which affect data consistency, doesn't that indicate you have database architecture issues? 
I think you won't be able to convince Doctrine contributors to change their opinion about `ENUM` unless you take on the [monumental task](https://github.com/doctrine/dbal/issues/6465#issuecomment-2214424147) of implementing `ENUM` support for all the database systems Doctrine works with.

**Comment by PowerKiKi** [comment]
You are correct, Doctrine won't change their opinion, and that's fine by me. However they could make it possible to implement it with custom code, and they should definitely strive not to break use-cases, as long as the maintenance cost is reasonable.

At best, dropping comments generated a lot of confusion, or at worse it broke some use-cases as illustrated by the amount of activity around that topic:

- https://github.com/doctrine/dbal/issues/6443
- https://github.com/doctrine/dbal/issues/6465
- https://github.com/doctrine/dbal/issues/6466
- https://github.com/doctrine/dbal/pull/6444
- https://github.com/doctrine/dbal/pull/6467
- https://github.com/doctrine/migrations/issues/1434
- https://github.com/doctrine/migrations/issues/1435

And it is not only limited to enum, according to https://github.com/doctrine/dbal/issues/5308#issuecomment-1058391601

It is a bit frustrating because there have been people warning of the situation [very early](https://github.com/doctrine/dbal/issues/5194#issuecomment-1018799512).

**Comment by berkut1** [comment]
You mentioned my issue as well :) 
Actually, I welcome the new type definition system because it solved (at least for me) many non-obvious errors that `DC2Type` was hiding by disabling part of the code. I have more of a question about the remnants of `DC2Type`, such as `inet -> string`, causing DBAL3 to ignore the schema and assume that `INET = VARCHAR` and left `INET` in migrations. However, in DBAL4 the logic changed, but the `inet -> string` rule remained, which seems strange and will be a surprise for many when they find out that all INETs have become VARCHARs. It would be more appropriate to remove `inet -> string` (and other similar types) altogether so that an exception is thrown and people can create a custom type themselves.

However, I don't understand why you can't do something [similar](https://github.com/doctrine/migrations/issues/1441#issuecomment-2211525188) for `ENUM`. I don't see a reason why it shouldn't work, unless if you're using third-party packages.

**Comment by greg0ire** [comment]
> But both of those solutions will change the DB colum from ENUM to VARCHAR

:thinking: 

Then why does the cook book solution 1 mention enums? `columnDefinition: "ENUM('visible', 'invisible')"`

**Comment by PowerKiKi** [comment]
`columnDefinition: "ENUM('visible', 'invisible')"` will "not work", meaning that it will keep generating migrations. Below I force to update the DB structure, then immediately ask for a diff. This should always be empty, but it incorrectly think that my field is not the same:


```php
#[ORM\Column(type: "string", columnDefinition: "ENUM('visible', 'invisible')")]
private $myTestField;
```

```sh
$ ./bin/doctrine orm:schema-tool:update --force && ./bin/doctrine orm:schema-tool:update --dump-sql | grep task
 Updating database schema...

     11 queries were executed

                                                                                                                        
 [OK] Database schema updated successfully!                                                                             
                                                                                                                        

ALTER TABLE task CHANGE my_test_field my_test_field ENUM('visible', 'invisible');    <=== INCORRECT
```

So that solution has no advantage over solution 2, because both generate infinite migrations. And it is very difficult to scale, and possibly requires a lost of copy/paste, since PHP attributes cannot use expressions.

So if we are stuck with infinite migrations anyway, I might as well use solution 2, which is easy to use and scale much better.

So the only choices we have are:

- VARCHAR with non-validated DB data
- ENUM with infinite migrations

It's a rather annoying regression in both cases in my opinion

**Comment by greg0ire** [comment]
> columnDefinition: "ENUM('visible', 'invisible')" will "not work", meaning that it will keep generating migrations. 

Weird, because I stumbled upon https://github.com/doctrine/dbal/pull/5224 recently, which is supposed to prevent that.

**Comment by berkut1** [comment]
> Weird, because I stumbled upon https://github.com/doctrine/dbal/pull/5224 recently, which is supposed to prevent that.

I suspect it's the same problem I tried to describe [here ](https://github.com/doctrine/dbal/issues/6466)using `INET` as an example. Apparently, I'm not very good at explaining things :)

**Comment by berkut1** [comment]
@PowerKiKi @michnovka 
If you are using ORM, could you test a theory? Remove this check (or only 464 line) and see if it helps with ENUM.
https://github.com/doctrine/orm/blob/9b09cd03c0f4577174dbf7af92644fa8f78faf64/src/Tools/SchemaTool.php#L463-L465

Especially if you are using this solution - https://www.doctrine-project.org/projects/doctrine-orm/en/3.2/cookbook/mysql-enums.html#solution-1-mapping-to-varchars

**Comment by meiyasan** [comment]
@berkut1 @greg0ire I have been facing similar problem in dbal#6444 ; using DBAL/ORM last released version.
I just quickly dug out a project implementing a custom EnumType class declaring a string column and using ENUM() declaration in db. I just tried to remove the line you mentioned and the migration issue still occurs.

I still can't get back the proper object neither. As far as my investigation went, this might perhaps work for pure PHP builtin enums; but still custom types would always be failling in my opinion: such as set types, or any other custom type requiring COMMENT information to translate back. I cannot see any other solution but making use of COMMENT and `DC2Type` declaration at the moment, but I am not an expert. Please advice.

**Comment by berkut1** [comment]
@xkzl what diff migration did you get after removing the line 464?

**Comment by PowerKiKi** [comment]
> @PowerKiKi @michnovka If you are using ORM, could you test a theory? Remove this check (or only 464 line) and see if it helps with ENUM. https://github.com/doctrine/orm/blob/9b09cd03c0f4577174dbf7af92644fa8f78faf64/src/Tools/SchemaTool.php#L463-L465
> 
> Especially if you are using this solution - https://www.doctrine-project.org/projects/doctrine-orm/en/3.2/cookbook/mysql-enums.html#solution-1-mapping-to-varchars

@berkut1, I tried that too. My mapping is not valid anymore, and I get the exception:


```
Doctrine\DBAL\Exception\InvalidColumnDeclaration: Column "role" has invalid type
[...]

Caused by
Doctrine\DBAL\Exception\InvalidColumnType\ColumnLengthRequired: Doctrine\DBAL\Platforms\MariaDB1060Platform requires the length of a VARCHAR column to be specified
```

**Comment by berkut1** [comment]
@PowerKiKi I see, different platforms have their own logic here. In your case, it doesn’t make sense to remove this check. My idea was that when checking the SQL declaration, the same values should be compared.

https://github.com/doctrine/dbal/blob/4.0.x/src/Platforms/AbstractPlatform.php#L2199-L2202

You can read more about it here. https://github.com/doctrine/dbal/issues/6468#issuecomment-2242324006

As I see in the first message of this issue, the migration function `down()` represents ENUM as `VARCHAR(0)`, so I assumed that maybe by default, DBAL sets the string length as 0 for MySQL/MariaDB and expects the same from ORM, but ORM assigns a standard length of 255 to any strings.

Overall, take a look at which strings are being compared here and how DBAL represent `ENUM` there:
https://github.com/doctrine/dbal/blob/4.0.x/src/Platforms/AbstractPlatform.php#L2199-L2202

**Comment by PowerKiKi** [comment]
I setup https://github.com/PowerKiKi/doctrine-migrations-issues-1441, a repository demonstrating that none of the approaches suggested throughout the entire discussion are working anymore. Feel free to create PR over there to test new approaches.

@berkut1 (and to a lesser extent @greg0ire), I think you are missing the point that DBAL issue is before the comparison step, as you suggested by mentioning `AbstractPlatform::columnsEqual()`. The issue is when we read the database structure. DBAL will introspect the DB and find `"enum('visitor','member','admin')"`. But then it will [incorrectly truncate the type down](https://github.com/doctrine/dbal/blob/389230389ed2d73a5fbf9a07d2bc0f8d0d56070e/src/Schema/MySQLSchemaManager.php#L120-L122) to only `"enum"`. So DBAL effectively assumes that `"enum('visitor','member','admin')"` and `"enum('new','active','archived')"` are the same type. But they clearly are not the same and are absolutely not interchangeable.

From that incorrect assumption it all goes downhill and [`$this->platform->getDoctrineTypeMapping($dbType)`](https://github.com/doctrine/dbal/blob/389230389ed2d73a5fbf9a07d2bc0f8d0d56070e/src/Schema/MySQLSchemaManager.php#L135) will incorrectly return the first (or last? :shrug:) custom PHP type that declared it is mapping to `"enum"` via its `\Doctrine\DBAL\Types\Type::getMappedDatabaseTypes()`.

The core issue really is that: **DBAL introspection incorrectly simplifies database types and mixes up unrelated database types**. Then of course custom PHP types cannot be reliable, if the foundation is broken. We should be able to use `getMappedDatabaseTypes()` to declare that the type maps to a specific enum, eg: `"enum('visitor','member','admin')"`. But we cannot.

In DBAL < 4.0 we were able to workaround that by using column comment in database, because it essentially completely bypassed auto-detection by "hardcoding" which custom PHP type it matches.

Since that was dropped, we don't have any way to keep the behavior pre-4.0. I am only paraphrasing https://github.com/doctrine/migrations/issues/1441#issuecomment-2207348574 here, until here.

I feel we have two solutions going forward:

- we fix `MySQLSchemaManager::_getPortableTableColumnDefinition()` to avoid truncating database types, and allowing proper type mapping via `getMappedDatabaseTypes()` declaration
- we re-introduce column comments to keep working around the historic issue

IMHO, the first solution is the proper one, but it also means it might break a lot more existing code. At least all enums and all sets and possibility other things too (depending on how well we code :sweat_smile:).

The second solution is the easiest and safest one. Restoring the code, and restoring concepts that everybody knows, and unbreaking existing code without any changes in user code. That might be less proper, but I'm inclined to still opt for that, because of the immediate simplicity and benefits.

All of this also applies to [SET](https://mariadb.com/kb/en/set-data-type/). And whichever solution we pick should also work for SET.

tl;dr: `"enum('visitor','member','admin')"` **!=** `"enum('new','active','archived')"`

**Comment by berkut1** [comment]
@PowerKiKi Okay, now I understand your point of view. It’s true that it doesn’t check the actual `ENUM` values, but for many custom types, this was also how it worked in DBAL3. In fact, it never really worked properly, and people used various hacks to get around it, like manually editing migrations.

It might be better to completely remove `ENUM` from DBAL, along with other similar types that no longer work because DBAL and ORM have different views on the default VARCHAR value, and because comment hints are no longer available.

**Comment by PowerKiKi** [comment]
> remove `ENUM` from DBAL

`ENUM` does not _exist_ in DBAL. There is no code that specifically mention that (except one single occurence in test). So there is nothing to remove really.

But I am a bit shocked that you suggested it. We've been asking to keep a way for us to implement our own support for ENUM, and you suggest the exact opposite. I feel I am not being heard at all here. I get it that DBAL has no interest to support it out-of-the-box. That's fair. But it feels like you are actively trying to prevent us to achieve our goal, whereas there could be a reasonable compromise to be made, but you don't seem to consider them seriously at all.

I would like to hear the opinion of other people too. Someone with much more insights that the both of us, such as @greg0ire or @derrabus ? would you accept a PR that avoid truncating type declaration for ENUM and SET ?

**Comment by berkut1** [comment]
>But I am a bit shocked that you suggested it. We've been asking to keep a way for us to implement our own support for ENUM, and you suggest the exact opposite

@PowerKiKi Not exactly. As far as I understand, the main contributors are strongly against bringing back comment hints, or something similar. If they're against it, then why keep non-functional pieces of code at all?

**Comment by michnovka** [comment]
@berkut1 you cannot remove `ENUM` from DBAL4 bacause it simply is not there. What you can remove is an example on docs for how to make custom ENUM() column. While it still works to create the DBAL schema, it will fail with migrations.

@PowerKiKi I think this thread has the underlying issue [described to the very detail](https://github.com/doctrine/migrations/issues/1441#issuecomment-2207027385):

> The issue is that oldSchema is created based on database ONLY. It ignores PHP code. And thats good, as the php code defines the new schema, if it took the PHP code as a hint of the type, then migration has no point, as it would be the same type always.

Any type which is not natively supported by DBAL will face this issue. We have a one-way function from PHP code -> DB representation, and we have no way to reverse this with accuracy, because

1. multiple PHP representations can lead to the same DBAL type
2. only default built-in DBAL types can be compared in migrations

With ENUMs specifically it is further complicated by the fact that basically every `ENUM` is a different type. DBAL was not meant to handle such case. It can handle `VARCHAR(16)` != `VARCHAR(63)` but this is because length is a valid property of `Column`. The list of valid properties is:

- name
- length
- precision
- scale
- unsigned
- fixed
- notnull
- default
- autoincrement

There is no way to specify `ENUM` cases. So the best we could get with current DBAL is `ENUM(1,2,3)` = `ENUM('a','b','c')`.

But that still assumes that `ENUM` would be a valid type in DBAL, which it is not (and there are no plans to make it so, because it is not abstract enough. Many DBMS do not support this. Its really a MySQL-family thing)

I do understand the reluctancy to bring back comments, but when the decision was made, I do not believe enough thought was given to the real implications. This is not just ENUMs. This is ANY nonstandard DBAL types. INET, SET etc. Exotic ones, yes, but 4.x breaks them.

While I agree to discourage from using typehints, I think this option should be brought back for the few cases, where without them 4.x is broken without a fix where 3.x worked just fine.

**Comment by derrabus** [comment]
Maybe we need some kind of verbatim type that all unknown database types are being mapped to for the sole purpose of making two introspections comparable.

**Comment by stof** [comment]
@derrabus as long as `AbstractSchemaManager::getPortableTableColumnDefinition` does not allow to access the original definition for enum fields during introspection, this verbatim type won't solve it.

**Comment by derrabus** [comment]
Sure that needs to be solved along with it.

**Comment by PowerKiKi** [comment]
> kind of verbatim type

It sounds like a way to say to DBAL "stop your magic, I'll handle everything from here". That's exactly what comment did. I suppose we can re-create the same concept with a different API. As long as I can use that new hypothetical API to implement a custom type to support ENUM and SET, I'd be happy. But at the same time, I can't help to wonder: why not just restore comments ? its code is simple, the concept is well-known and proven to work even (especially!) from user code.

The only rationale we were given for comment removal was "DBAL core does not need it anymore". But I think we've proven that there are reasonable use-cases, implemented outside DBAL core, that requires it. What is the reason to block restoring comments ?

**Comment by stof** [comment]
@PowerKiKi comments were solving some cases, and were *creating* the same kind of issue (endless change) in *other* cases.
this verbatim type used during introspection to make it use the actual unmodified column declaration (which is then compared to the one generated by your custom type) looks like the proper fix for the root cause of the issue (solution 1 in https://github.com/doctrine/migrations/issues/1441#issuecomment-2285460448)

**Comment by greg0ire** [comment]
> as long as AbstractSchemaManager::getPortableTableColumnDefinition does not allow to access the original definition for enum fields during introspection, this verbatim type won't solve it.

@stof @derrabus  should we use `Doctrine\DBAL\Schema::$_columnDefinition` for this? Or is it a bad idea/meant for something else?

**Comment by PowerKiKi** [comment]
@derrabus, I'd be willing to help introduce a "verbatim type", but I'm not quite sure what you mean codewise. Would you be able to write some pseudo-code to help me understand ? or would you prefer to prototype it yourself ?

**Comment by WubbleWobble** [comment]
Running into similar issues after upgrading.

We had a DateTimeMicroseconds type as based upon various code snippets in <https://github.com/doctrine/dbal/issues/2873>, but instead of replacing the datetime type completely, we complemented it so that we could have some columns that were just seconds-based, and other columns that supported microseconds.

Now that DBAL doesn't support comments, this behaviour has been broken.

`platform->getDoctrineTypeMapping(dbType)` seems to base its decision purely on `DATETIME`, so I don't think I can get it to return a different type based upon the column length.

Similarly `type->convertToDatabaseValue` and `type->convertToPHPValue` don't provide column meta, so I don't think I can see the column size in order to use a different datetime format string depending upon the column length.

I could try to override `AbstractPlatform->getDateTimeFormatString` to add ".u", but the `type->convertToPHPValue` would break if provided a non-microsecond string.

The remaining potential options I have identified are:
a) Go back to DBAL 3.x which supports comments
b) Override datetime to overwrite the dateTimeFormat string used, and switch all of my columns over the DATETIME(6) - i.e. no more DATETIME(0)
c) Override datetime and add logic to `type->convertToPHPValue` to inspect whether the string coming out has a ".123456" microsecond part, and then using either the normal or microsecond dateTimeFormat string based upon that.

The comments stuff was useful. 

In my particular case, I could make the type decision if I had access to more column meta - via hypothetical method changes such as `getDoctrineTypeMapping($dbType, $columnMeta)` or `convertToDatabaseValue($value, $platform, $columnMeta)`

(N.B. Not a Doctrine expert - I only know what I've gleamed from reading comments here and digging through the code. If I'm missing the elephant in the room and what I'm trying to do is actually possible, please let me know! :D)

**Comment by derrabus** [comment]
> @derrabus, I'd be willing to help introduce a "verbatim type", but I'm not quite sure what you mean codewise. Would you be able to write some pseudo-code to help me understand ? or would you prefer to prototype it yourself ?

If you're motivated to work on this, please go ahead. I will support you if I can.

The idea is that during introspection DBAL maps column to this verbatim type if it cannot map it to a registered DBAL type. And since your custom enum type and the verbatim type should ideally generate the same SQL, no migration is generated.

---

> We had a DateTimeMicroseconds type

Multiple attempts have been made to support `DATETIME` fields with configurable precision natively in DBAL. In your case, I'd rather try to finish that feature. This feature is supported by most DBMS and you're not the only one who needs this. Instead of wasting time on documenting workarounds, we should really focus on getting that implemented properly.

**Comment by WubbleWobble** [comment]
> Multiple attempts have been made to support `DATETIME` fields with configurable precision natively in DBAL. In your case, I'd rather try to finish that feature. This feature is supported by most DBMS and you're not the only one who needs this. Instead of wasting time on documenting workarounds, we should really focus on getting that implemented properly.

Well - that's kind one of the angles I was looking at - having the DateTimeType react appropriately depending upon the precision, but `convertToPHPValue` and `convertToDatabaseValue` don't have access to the column information and so can't change their behaviour based upon it.

The next hop up from that was where types are determined - i.e. `getDoctrineTypeMapping(string $dbType)` - and that too didn't have the information (i.e. it only knew "datetime" and as such gave out a DateTimeType), so it wasn't possible there either as far as I could see.

So my understanding ended up being that implementing it would require a change in the API / interfaces.

**Comment by berkut1** [comment]
>@PowerKiKi I'd be willing to help introduce a "verbatim type", but I'm not quite sure what you mean codewise. Would you be able to write some pseudo-code to help me understand ? or would you prefer to prototype it yourself ?

I think we need to introduce a new type that will just store the database data type as it is. For example, right now, if a type like `enum('apple','orange','pear') `comes in, then here:

https://github.com/doctrine/dbal/blob/4.1.x/src/Schema/MySQLSchemaManager.php#L121

It turns into just `enum`, and we permanently lose the details of the enum. We need to keep the full `enum('apple','orange','pear')`, possibly in a new property `Doctrine\DBAL\Schema::$_ColumnDefinitionAsItIs` (we probably can't use old `$_columnDefinition`, because of https://github.com/doctrine/dbal/blob/a41d81d7d255d4f855606b4a29c792e062acb554/src/Platforms/AbstractPlatform.php#L2222-L2223) and also let Doctrine know that this type will be verbatim. For example, register it like this:

```php
$conn->getDatabasePlatform()->registerDoctrineTypeMapping('enum', 'verbatim');
```

And then in the custom type, just do something like this:

```php
class EnumType extends VerbatimType
{
    public function getSQLDeclaration(array $column, AbstractPlatform $platform): string
    {
        return "enum('apple','orange','pear')";
    }
}
```

VerbatimType probably will looks like this:

```php
class VerbatimType extends Type
{
    public function getSQLDeclaration(array $column, AbstractPlatform $platform): string
    {
        return $column['сolumnDefinitionAsItIs']; //this seems not right :)
    }
}
```

However, we need to make it work for any unknown types, and this might require a lot of research on how different DBMSs store data types that are unknown to Doctrine, and whether we can simply save all of them in a `сolumnDefinitionAsItIs`.

P.S Of course, my idea might be wrong.

**UPD:**
Probably, a verbatim type will even allow us to skip the type registration step. For example, right now, simple unknown types that don't contain anything except the TYPE_NAME (like in PostgreSQL with `INET`, `CIDR`, etc), meaning without `()`, need to be registered like this:

```yaml
doctrine:
    dbal:
        mapping_types:
            inet: alias_name

        types:
            alias_name: { class: 'App\Entity\InetType' }
```

Maybe, with a verbatim type, we can skip `mapping_types` if Doctrine automatically marks all unknown types as verbatim (currently, Doctrine throws an exception or registers as `string`).

**Comment by tomsykes** [comment]
Setting the Enum case aside for a moment (as I think it is muddying the water WRT what I think is the underlying issue)…

As of DBAL 4, the representation in the database is no longer an exact match for the schema definition from code, so when the comparator makes its comparison, there can be a difference even when nothing has changed.

---

Take the example of a custom type that is an RGB hex value represented by an object with 3 integer properties.
```php
class RGB {
    public int $red;
    public int $green;
    public int $blue;
}
```

This is mapped to a varchar in the database using a custom type: e.g. `{red: 255, green: 255, blue: 255}` => `FFFFFF`

At runtime, DBAL correctly maps the varchar data from the database to my custom RGB object using `App\DBAL\Types\RGBType` (and vice-versa), because it is using the schema def from code - no problem.

When using the comparator to look for schema updates:
- loads the schema from the database, and since the column type is varchar, resolves that to `Doctrine\DBAL\Types\StringType`
- loads the schema from code (which maps the column to `App\DBAL\Types\RGBType`)
- they're different - problem

Prior to the removal of database comments, the first step above would have resolved the column to the correct type.

**Comment by derrabus** [comment]
> When using the comparator to look for schema updates:
> 
> * loads the schema from the database, and since the column type is varchar, resolves that to `Doctrine\DBAL\Types\StringType` 
> * loads the schema from code (which maps the column to `App\DBAL\Types\RGBType`)
> * they're different - problem

No. As long as both types generate the same DDL statements, they are to be be considered equal.

**Comment by tomsykes** [comment]
> No. As long as both types generate the same DDL statements, they are to be be considered equal.

But the comparator isn't comparing DDL statements, it's comparing `Doctrine\DBAL\Schema\Column` definitions, and the `type` property in `Column` is a resolved DBAL type, not a DDL type.

![image](https://github.com/user-attachments/assets/e614b639-2ab1-4511-a831-f70ba4620565)
![image](https://github.com/user-attachments/assets/109897f3-307b-43c0-9b37-777149bdfc6d)

i.e. it's Doctrine\DBAL\Types\StringType and not "VARCHAR"

if this were comparing DDL statements, it would be working fine

**Comment by WubbleWobble** [comment]
> But the comparator isn't comparing DDL statements, it's comparing `Doctrine\DBAL\Schema\Column` definitions, and the `type` property in `Column` is a resolved DBAL type, not a DDL type.

This is exactly the cause of my problems. My custom type works fine in use, but the comparator thinks that something has changed and causes schema-update  / migration diff to perpetually recommend an "ALTER TABLE" on said column.

![image](https://github.com/user-attachments/assets/70ede352-130f-4852-b9c6-20e3dc584bc9)

Here I've dumped the $oldSchema and $newSchema when running `doctrine:schema:update --dump-sql`.

It thinks they are different because it has two different DBAL types, whereas if it compared raw column types, it'd realise that the schema wants a DATETIME, and the database already has a DATETIME, and all would be fine! :)

**Comment by derrabus** [comment]
> But the comparator isn't comparing DDL statements

Yes, it does. 

https://github.com/doctrine/dbal/blob/a41d81d7d255d4f855606b4a29c792e062acb554/src/Platforms/AbstractPlatform.php#L2225-L2230

Only if two columns are considered to be not equal, a `ColumnDiff` object is created which contains the method from your screenshot.

https://github.com/doctrine/dbal/blob/a41d81d7d255d4f855606b4a29c792e062acb554/src/Schema/Comparator.php#L183-L187

**Comment by WubbleWobble** [comment]
Thanks @derrabus - that gives me somewhere further to dig to try to understand what is going on! :)

**Comment by WubbleWobble** [comment]
> > But the comparator isn't comparing DDL statements
> 
> Yes, it does.

I've looked into this further - thanks for the pointer! - and it's comparing the DDL statements generated based by the resolved Doctrine DBAL types.

https://github.com/doctrine/dbal/blob/a41d81d7d255d4f855606b4a29c792e062acb554/src/Platforms/AbstractPlatform.php#L1392

So essentially in my case:
a) It's calling `DateTimeType->getSQLDeclaration()` and the results of `getColumnDeclarationSQL()` end up as being `DATETIME NOT NULL` (old schema)
b) It's calling `DateTimeMicrosecondType->getSQLDeclaration()` and the results of `getColumnDeclarationSQL()` end up as being `DATETIME(6) NOT NULL` (new schema)

It's then raising a ColumnDiff as you suggested.

However, what is actually in the database is indeed a `DATETIME(6) NOT NULL`, so the DDL generated for "old schema" is not actually what is in the database.

Ultimately I've tracked this down to [`AbstractMysqlPlatform->getDateTimeDeclarationSQL()`](https://github.com/doctrine/dbal/blob/4.1.x/src/Platforms/AbstractMySQLPlatform.php#L164-L171) which doesn't seem to care about the [fsp value](https://dev.mysql.com/doc/refman/8.4/en/fractional-seconds.html) on DATETIME - I declare a "length" (fsp) of 6, but it's not using it:

```php
    public function getDateTimeTypeDeclarationSQL(array $column): string
    {
        if (isset($column['version']) && $column['version'] === true) {
            return 'TIMESTAMP';
        }

        return 'DATETIME';
    }
```

It therefore seems I could override this method to resolve my comparator woes?

I could also make a quick patch for this issue - it looks pretty trivial - if this sounds correct to you?
(i.e. For all non-null/zero column "lengths", append that "length" - e.g. DATETIME(3) or TIMESTAMP(6). Valid values for MySQL are 0-6. I would need to check whether this "fsp" has always been supported on MySQL and if not, when it was introduced etc)

-----

Also it's a bit odd, in that it's not comparing the actual old column definition, but rather looking at the old column definition, parsing it into bits, deciding upon a DBAL Type class, and then asking that Type class to generate what it thinks the column definition should look like based upon those parsed bits, and then using *that*. I guess there are nuanced reasons to do that, but on the face of it, it does seem rather around the houses.

**Comment by derrabus** [comment]
> It therefore seems I could override this method to resolve my comparator woes?

That could probably work.

> I could also make a quick patch for this issue - it looks pretty trivial - if this sounds correct to you? (i.e. For all non-null/zero column "lengths", append that "length" - e.g. DATETIME(3) or TIMESTAMP(6). Valid values for MySQL are 0-6. I would need to check whether this "fsp" has always been supported on MySQL and if not, when it was introduced etc)

You can have a look at the previous attempt (doctrine/dbal#5961) and try to complete it. The workaround that you need might be trivial. However, properly supporting DATETIME with variable precision is not. You could also open an issue on the DBAL repository first where we discuss the obstacles that you might run into.

**Comment by WubbleWobble** [comment]
> You can have a look at the previous attempt ([doctrine/dbal#5961](https://github.com/doctrine/dbal/pull/5961)) and try to complete it

I've looked at it, and it looks like the original author has put a lot of work into this and as far as I can see has resolved every change request.

I can see the entitled muppet (that's the polite form) throwing a spanner in the works there with his rudeness, but what further works needs to be done to it / how can people help to progress it?

**Comment by derrabus** [comment]
It has to be ported to the 4.2 branch at least. A lot of time has passed since then, I don't recall the remaining blockers right now. But please, let's discuss this on a separate issue on the DBAL repository. This is getting more and more off-topic.

**Comment by berkut1** [comment]
@derrabus What do you think about this idea for implementing the verbatim type https://github.com/doctrine/migrations/issues/1441#issuecomment-2295517648?  Or do you have a different approach in mind?

**Comment by derrabus** [comment]
@berkut1 There should be no need to register or extend that verbatim type. It is used only during introspection, if a DB column type is encountered that DBAL has no mapping for.

**Comment by b3n3d1k7** [comment]
So solution 2 described here: 
https://www.doctrine-project.org/projects/doctrine-orm/en/3.2/cookbook/mysql-enums.html#solution-2-defining-a-type
does not work anymore or am I missing something?
This leads to infinite migrations and there is no workaround, right?
![image](https://github.com/user-attachments/assets/27c286f3-8ace-40b1-83da-1a2c90441ab9)

**Comment by derrabus** [comment]
@b3n3d1k7 No, it doesn't. Please read the full thread before commenting, sorry.

**Comment by uncaught** [comment]
We've had this issue for ages with our own custom enums and solved it with a patch for MySQLSchemaManager.php:

```diff
diff --git a/src/Schema/MySQLSchemaManager.php b/src/Schema/MySQLSchemaManager.php
index bd383b012..556e3df4e 100644
--- a/src/Schema/MySQLSchemaManager.php
+++ b/src/Schema/MySQLSchemaManager.php
@@ -278,6 +278,8 @@ class MySQLSchemaManager extends AbstractSchemaManager
 
         $column = new Column($tableColumn['field'], Type::getType($type), $options);
 
+        $column->setPlatformOption('hcCurrentTableColumnType', $tableColumn['type']);
+
         if (isset($tableColumn['characterset'])) {
             $column->setPlatformOption('charset', $tableColumn['characterset']);
         }
```

Then in our type we simply do this to return the actual current SQL declaration for comparism:

```php
  public function getSQLDeclaration(array $column, AbstractPlatform $platform): string {
    // This is patched at the end of MySQLSchemaManager::_getPortableTableColumnDefinition to allow diffs on enums:
    if (isset($column['hcCurrentTableColumnType'])) {
      return $column['hcCurrentTableColumnType'];
    }
    //... return your `enum(...)`, but this must be lower case for comparism!
```


In case this helps anyone.

**Comment by michnovka** [comment]
has anybody started work on the Verbatim type? I might have some time in end of October/November to look into this.

**Comment by PowerKiKi** [comment]
I thought I would give it a try, but I didn't yet. And I won't have much time for a while. Feel free to start whenever you want. And notice everybody here when you do, so that we don't duplicate work.

**Comment by morozov** [comment]
Linking a couple more existing issues for reference: https://github.com/doctrine/dbal/issues/4470, https://github.com/doctrine/dbal/issues/5306.

**Comment by VincentLanglet** [comment]
I encounter the same issue with my custom types.
I keep generating dozen of lines like
```
$this->addSql('ALTER TABLE activity_events CHANGE status status ENUM(\'DISPATCHED\', \'ONGOING\', \'SUCCEEDED\', \'FAILED\') NOT NULL');
```
in my migration diff, when it worked fine with `COMMENT \'(DC2Type:enumActivityEventStatus)\''`

It does not give a good experience with the DBAL 4 migration, and it seems like it impacts a lot of developer.

I read all the discussion and maybe I miss-understand but I feel like there is currently no solution to avoid this extra sql in every diff. Isn't it ?

The comment was really useful for custom type. If there is no known solution, what about re-introducing it back ? At least this could be optional like an option to toggle. This way if comments have bad impact for some user they can disable it.

**Comment by uncaught** [comment]
As I said, we have had a solution for years and it kept working with dbal 4.

I would create a PR, but it's for mysql only 🫤

**Comment by meiyasan** [comment]
> I encounter the same issue with my custom types.
> 
> I keep generating dozen of lines like
> 
> ```
> 
> $this->addSql('ALTER TABLE activity_events CHANGE status status ENUM(\'DISPATCHED\', \'ONGOING\', \'SUCCEEDED\', \'FAILED\') NOT NULL');
> 
> ```
> 
> in my migration diff, when it worked fine with `COMMENT \'(DC2Type:enumActivityEventStatus)\''`
> 
> 
> 
> It does not give a good experience with the DBAL 4 migration, and it seems like it impacts a lot of developer.
> 
> 
> 
> I read all the discussion and maybe I miss-understand but I feel like there is currently no solution to avoid this extra sql in every diff. Isn't it ?
> 
> 
> 
> The comment was really useful for custom type. If there is no known solution, what about re-introducing it back ? At least this could be optional like an option to toggle. This way if comments have bad impact for some user they can disable it.

@VincentLanglet Your feeling is correct, in my opinion. I haven’t found any alternatives either. I’ve been begging for merging this PR https://github.com/doctrine/dbal/pull/6444 for months, but they don’t want to for obscure reasons. 
While this doesn’t have to be a permanent solution, removing those lines too early feels like a quick shot. They could just leave them in place until a more mature alternative is found.

A good legacy solution would be to follow the approach of the doctrine/annotation bundle and make it optional, but they seem to have missed the point for months.

If you’re interested, I created this package - a code modifier that reintroduces the missing lines: https://packagist.org/packages/glitchr/doctrine-dc2type. Please use with caution and inspect the package on a fresh symfony skeleton if you wish, there is nothing too complex or fancy, but it is nothing official.
This is the only alternative I had for the sake of my projects.

**Comment by derrabus** [comment]
Your complaints have been duly noted. Now, can we please get back to discussing solutions?

I mean, you can just stay on DBAL 3 until the problem is solved or help working on a proper solution. We will still maintain DBAL 3 for some time.

**Comment by VincentLanglet** [comment]
> Now, can we please get back to discussing solutions?

Sure, but I'm unsure how to help. I certainly lack of knowledge about how this work.

If I debug `MySQLSchemaManager::_getPortableTableColumnDefinition`, I feel like the logic here is incorrect
https://github.com/doctrine/dbal/blob/7a8252418689feb860ea8dfeab66d64a56a64df8/src/Schema/MySQLSchemaManager.php#L118-L135

When the data_type in my DB is `enum('DISPATCHED','ONGOING','SUCCEEDED','FAILED')`
- `$dbType = strtok($dbType, '(), ')` is `enum`
- `$length = $tableColumn['length'] ?? strtok('(), ');` is `'dispatched`

Personally, I could almost fix my issue if I overriden the 
```
public function getMappedDatabaseTypes(AbstractPlatform $platform): array
```
method in my customs Type, but it would require to update the MySQLSchemaManager to something like
```
        $dbType = strtolower($tableColumn['type']);
        if (!str_starts_with($dbType, 'enum(')) {
            $dbType = strtok($dbType, '(), ');
            assert(is_string($dbType));

            $length = $tableColumn['length'] ?? strtok('(), ');
        } else {
            $length = null;
        }
```

Then https://www.doctrine-project.org/projects/doctrine-orm/en/latest/cookbook/mysql-enums.html could be updated to implement
```
public function getMappedDatabaseTypes(AbstractPlatform $platform): array
    {
        $values = array_map(function($val) { return "'".$val."'"; }, $this->values);

        return ["ENUM(".implode(",", $values).")"];
    }
```
(the abstract EnumType could even be provided by doctrine eventually...) 

Do you think it could a good improvement @greg0ire @derrabus ? (And how avoid a BC break...)
 
> I mean, you can just stay on DBAL 3 until the problem is solved or help working on a proper solution. We will still maintain DBAL 3 for some time.

DBAL 4 still have benefits ; I like the evolution done to bigint.

**Comment by michnovka** [comment]
Should the `VerbatimType` live inside Migration or DBAL? I would argue Migration, since it should not be used anywhere else.

And @stof 
> as long as AbstractSchemaManager::getPortableTableColumnDefinition does not allow to access the original definition for enum fields during introspection, this verbatim type won't solve it.

This is not true, we have that, see https://github.com/doctrine/migrations/issues/1441#issuecomment-2206338070

We can just create a `VerbatimType` from within here and just give it this `$tableColumn` property. We cannot use TypeRegistry for this case, as every `VerbatimType` will be its own instance.

But I think we can handle all this inside `SchemaManager`'s `_getPortableTableColumnDefinition()`. Seems well have to edit every single one tho, which overrides this method and introduce this logic.

Then we just make special case in comparator for `VerbatimType` and we're golden.

Am I missing something, or does it sound too easy to you too?

**Comment by derrabus** [comment]
Does doctrine/dbal#6536 solve the issue for those of you who are using enum columns?

**Comment by stof** [comment]
@michnovka both the schema introspection and the schema comparison are provided by DBAL. So it would definitely be used outside Migrations (and Migrations would not even use it directly but only through the usages in DBAL itself)

**Comment by VincentLanglet** [comment]
> Does [doctrine/dbal#6536](https://github.com/doctrine/dbal/pull/6536) solve the issue for those of you who are using enum columns?

I'm getting `No changes detected in your mapping information.` with your branch when I run do:mi:diff.
That's an error message I like.

Also I tried to change my mapping and/or change my enumType ; in both case the diff is correct.
It's awesome, thanks a lot @derrabus

**Comment by michnovka** [comment]
Thanks @derrabus this indeed solves the issues with ENUM and migrations.

Only thing I noticed was that if my sqlDeclaration defined the type as lowercase `enum('a','b')` it would still cause migrations (where up would have `ENUM('a','b')` and down `enum('a','b')`. Is this expected?

Changing my custom type's sql declaration to uppercase ENUM resolved this.

**Comment by derrabus** [comment]
So, this is your fix then. 🙂

**Comment by PowerKiKi** [comment]
@derrabus (or @michnovka ?), could you please clarify how the new `EnumType` should be used ?

I updated [my playground](https://github.com/PowerKiKi/doctrine-migrations-issues-1441) to the latest versions of all packages, and I [still see `varchar` in my DB](https://github.com/PowerKiKi/doctrine-migrations-issues-1441/actions/runs/11363309103/job/31606983791#step:8:41), instead of the expected `enum`. I thought it should work out of the box. Did I misunderstand something ?

The [model is declared there](https://github.com/PowerKiKi/doctrine-migrations-issues-1441/blob/master/src/Cases/PropertyMapping/Model/User.php), and [Doctrine is configured](https://github.com/PowerKiKi/doctrine-migrations-issues-1441/blob/ea53e65de1ef14df61b849b0255eec0e0534dddf/tests/AllTest.php#L77-L101) in the simplest manner I know of.

Basically I am doing:

```php
<?php

use Doctrine\ORM\Mapping as ORM;

enum UserRole: string
{
    case Visitor = 'visitor';
    case Member = 'member';
    case Admin = 'admin';
}


enum UserStatus: string
{
    case New = 'new';
    case Active = 'active';
    case Archived = 'archived';
}

#[ORM\Entity]
class User
{
    #[ORM\Id]
    #[ORM\Column(type: 'integer')]
    public ?int $id = null;

    #[ORM\Column]
    public UserRole $role = UserRole::Visitor;

    #[ORM\Column]
    public UserStatus $status = UserStatus::New;
}
```

And that:

```php
function createEntityManager(): EntityManager
{
    $config = ORMSetup::createAttributeMetadataConfiguration(['my-path-to-my-entity'], true);

    $connection = DriverManager::getConnection([
        'driverClass' => Driver::class,
        'host' => '127.0.0.1',
        'dbname' => 'test',
        'user' => 'test',
        'password' => 'test',
    ], $config);

    $entityManager = new EntityManager($connection, $config);

    return $entityManager;
}
```

...but it still fails to have `enum` in my DB.

**Comment by michnovka** [comment]
@PowerKiKi still by default doctrine will use `StringType` for your enum. You will need to customize

```php
    #[ORM\Column(type: 'enum', enumType: UserRole::class, values: [...])]
    public UserRole $role = UserRole::Visitor;
```

If you do not want to specify values here, and want by default all enum values to be used in DB as well, you have to use a custom type.

You can look into https://github.com/Elao/PhpEnums that makes this much easier.

You can also look into https://github.com/Elao/PhpEnums/issues/212 where I show on top my implementation which auto-registers enums as types.

**Comment by derrabus** [comment]
> @derrabus (or @michnovka ?), could you please clarify how the new `EnumType` should be used ?

Please read doctrine/orm#11666.

> If you do not want to specify values here, and want by default all enum values to be used in DB as well, you have to use a custom type.

No, don't do that. Custom enum types should be obsolete now.

> You can look into https://github.com/Elao/PhpEnums that makes this much easier.

Not anymore, really.

**Comment by michnovka** [comment]
@derrabus this is so cool, thanks a lot, I was unaware of this new PR

@PowerKiKi if you still want to use MySQL enums by default, you can take advantage of TypedFieldMapper. Here is my solution for the interested:

1. create custom `EnumTypedFieldMapper`
```php
<?php

declare(strict_types=1);

namespace App\Doctrine\ORM\TypedFieldMapper;

use Doctrine\DBAL\Types\Types;
use Doctrine\ORM\Mapping\TypedFieldMapper;
use ReflectionNamedType;
use ReflectionProperty;

class EnumTypedFieldMapper implements TypedFieldMapper
{
    /**
     * {@inheritdoc}
     */
    public function validateAndComplete(array $mapping, ReflectionProperty $field): array
    {
        $type = $field->getType();

        if (
            ! isset($mapping['type'])
            && ($type instanceof ReflectionNamedType)
        ) {
            if (!$type->isBuiltin() && enum_exists($type->getName())) {
                $mapping['type'] = Types::ENUM;
            }
        }

        return $mapping;
    }
}
```
2. adjust `services.yaml` to register this service and also pass it into the `ChainTypedFieldMapper` together with the `DefaultTypedFieldMapper` (note that in my example you can even see how I map some other typed fields to DBAL types)
```yaml
services:
    App\Doctrine\ORM\TypedFieldMapper\EnumTypedFieldMapper:

    Doctrine\ORM\Mapping\DefaultTypedFieldMapper:
        arguments:
            $typedFieldMappings:
                Decimal\Decimal: 'decimal'
                Carbon\CarbonImmutable: 'datetime_immutable'

    Doctrine\ORM\Mapping\ChainTypedFieldMapper:
        arguments:
            $typedFieldMappers:
                - '@App\Doctrine\ORM\TypedFieldMapper\EnumTypedFieldMapper'
                - '@Doctrine\ORM\Mapping\DefaultTypedFieldMapper'
```
3. then replace the default `DefaultTypedFieldMapper` with your newly configured `ChainTypedFieldMapper` inside `doctrine.yaml` - note that the chain makes sure that the default mapper is still used, just executed AFTER your custom enum mapper
```yaml
doctrine:
    orm:
        typed_field_mapper: 'Doctrine\ORM\Mapping\ChainTypedFieldMapper'
```

now you can use
```php
#[ORM\Column]
public UserRole $role = UserRole::Visitor;
```

And it will make your field an ENUM type in database

Note that this requires ORM 3.3, DBAL 4.2 and DoctrineBundle 2.13

**Comment by PowerKiKi** [comment]
@derrabus, this is great, thank you !

My playground is now [working as expected](https://github.com/PowerKiKi/doctrine-migrations-issues-1441/actions/runs/11370001327/job/31628766163#step:8:18). Specifying the type of the column is an obvious solution I should have thought about, `#[ORM\Column(type: 'enum')]` !

@michnovka, thanks for the hint for `ChainTypedFieldMapper`. It is a nice to have that I might implement at some point.

From my point of view, which is only concerned about enum, this issue may be closed as resolved.

**Comment by acasademont** [comment]
Well, we still have the issue with the `DATETIME(6)`  as @WubbleWobble pointed out in https://github.com/doctrine/migrations/issues/1441#issuecomment-2296926524

Not sure if this should remain open as a "placeholder" for that problem too.

A short patch that adds a middleware to the connection setting a custom platform (extended on `MysqlPlatform80`) and modifying the sql snippet to take into account the length. This now generates empty diffs for us, hope it might help someone else.

```
use Doctrine\DBAL\Driver;
use Doctrine\DBAL\Driver\Middleware;
use Doctrine\DBAL\Platforms\AbstractPlatform;
use Doctrine\DBAL\Platforms\MySQL80Platform;
use Doctrine\DBAL\ServerVersionProvider;

class LengthAwareMysqlMiddleware implements Middleware
{
    public function wrap(Driver $driver): Driver
    {
        return new class($driver) extends Middleware\AbstractDriverMiddleware {
            public function getDatabasePlatform(ServerVersionProvider $versionProvider): AbstractPlatform
            {
                return new class extends MySQL80Platform {
                    public function getDateTimeTypeDeclarationSQL(array $column): string
                    {
                        $sql = parent::getDateTimeTypeDeclarationSQL($column);

                        if (isset($column['length']) && $column['length'] > 0) {
                            $sql .= '(' . $column['length'] . ')';
                        }

                        return $sql;
                    }
                };
            }
        };
    }
}
```

**Comment by derrabus** [comment]
> Well, we still have the issue with the `DATETIME(6)`  as @WubbleWobble pointed out in https://github.com/doctrine/migrations/issues/1441#issuecomment-2296926524

… and my answer to his comment still applies.

**Comment by tomsykes** [comment]
I still generally have a problem with the dropping of D2Ctype comments. By doing so. Doctrine has lost the ability to have more than one DBAL type having the same column definition.

When the comparator tries to detect if anything has changed, it's using the column definition to determine the DBAL type. When there are two DBAL types with the same column definition, it just picks the first one. When there were D2Ctype comments, the comparator could distinguish between the two DBAL types.

**Comment by uncaught** [comment]
> When the comparator tries to detect if anything has changed, it's using the column definition to determine the DBAL type.

It does? I can't imagine this is the case. It would have to run through all possible parameters, too, to find the type based on a schema. Just imagine it finds a `decimal(42,7)` - how would that deterministically be resolved to a DBAL type?

All the migrations need to do is compare what is with what should be. And for that the comments were superfluous.

**Comment by tomsykes** [comment]
I think you're right @uncaught, you would expect the comparator to be generating a dummy schema based on the current state of the entities, and then compare that with the actual schema, but that's not what's happening.

@michnovka details how it works in their [first comment](https://github.com/doctrine/migrations/issues/1441#issuecomment-2206338070)

Notice the call to `$platform->getDoctrineTypeMapping($dbType)`. It's resolving a doctrine type based on the database definition.

Edit: lets be more precise about this... the comparator isn't generating anything, it's just comparing what it's being given... but the point is the the `AbstractSchemaManager` + `[PlatformSpecific]SchemaManager` are providing the comparator with column definitions that are based on a resolved DBAL type, and not the actual underlying SQL type, so when it comes to the comparison it's not necessarily a fair representation of what's actually in the database - hence why the migrations are being continually generated.

**Comment by Pixelshaped** [comment]
I upgraded all my enums from the legacy custom type (as documented) to the new enum type, but still get the migrations generated every time. I guess this is because the declaration is in uppercase but the db sends the type in lowercase.
@michnovka I feel like you encountered this issue, how did you solve it? Did you keep a custom declaration somewhere?

Edit: Not sure it's that: as a quick workaround I tweaked the vendor `AbstractMySQLPlatform::getEnumDeclarationSQL` to return `enum` instead of `ENUM` and still get the error.

Without tweaking anything my redundant migrations look like this:
```
public function up(Schema $schema): void {
    $this->addSql('ALTER TABLE address CHANGE country country ENUM(\'fr\', \'de\', \'nl\', \'pl\', \'gb\', \'be\') NOT NULL');
    // ...
}
```

```
public function down(Schema $schema): void {
    $this->addSql('ALTER TABLE address CHANGE country country VARCHAR(0) NOT NULL')
    // ...
}
```

Edit 2: when executing `doctrine:migrations:diff` I saw that the mapping for the "from" schema was enum string. Indeed in my conf I still had:
```
doctrine:
    dbal:
        connections:
            default:
                mapping_types:
                    enum: string
```
Removing it obviously solved the issue. Leaving this there in case it helps someone.

So: no issue, and thanks @derrabus!

**Comment by taleteller** [comment]
After migrating a project to Symfony 6.4 i stumbled upon this issue. While it is fixable for ENUM datatypes, it is still an issue with SET datatypes. While there is a `Doctrine\DBAL\EnumType` no similar `Doctrine\DBAL\SetType` exists. Like all custom types the underlying issue #1435 persists.
